[WIP] Higher-moments pens

n8willis opened the pull request:
Work-in-progress on adding the higher moments to MomentsPen and StatisticsPen.

Right now it only has 2nd and 4th Polar moments.

It's looking to me like the odd-numbered Polar moments are not doable without changing `symfont`.

As in the [discussion on 2339](https://github.com/fonttools/fonttools/issues/2339#issuecomment-871592041), the original integral is all hyperbolic & trig & logs and doesn't really simplify. There might be numerical approximations for those functions that would work, but likely it'd either mean changing GreenPen or adding another one. (I haven't actually tried; that's just the current state and what I think from poking at it.)

behdad commented:
I've moved the definition of OpenContourError to basePen. Can you update that please?

behdad commented:
Humm. This needs rebasing on top of main since i changed symfont and momentsPen to use Cython.

behdad commented:
I can look into it in a few days if you don't feel like it.

n8willis commented:
> Humm. This needs rebasing on top of main since i changed symfont and momentsPen to use Cython.

> I've moved the definition of OpenContourError to basePen. Can you update that please?

Ahh; sure thing. That's just a side-effect from me not checking back in so frequently. Presumably that applies to the existing MomentsPen PR, too.

behdad commented:
> > Humm. This needs rebasing on top of main since i changed symfont and momentsPen to use Cython.
> 
> > I've moved the definition of OpenContourError to basePen. Can you update that please?
> 
> Ahh; sure thing. That's just a side-effect from me not checking back in so frequently. Presumably that applies to the existing MomentsPen PR, too.

Yes.

n8willis commented:
> > > Humm. This needs rebasing on top of main since i changed symfont and momentsPen to use Cython.
> > 
> > 
> > > I've moved the definition of OpenContourError to basePen. Can you update that please?
> > 
> > 
> > Ahh; sure thing. That's just a side-effect from me not checking back in so frequently. Presumably that applies to the existing MomentsPen PR, too.
> 
> Yes.

Yeah; I'll try and do this tonight. Been occupied doing other absurd things like DFTs of glyphs (which thankfully will never end up an Pens....).

behdad commented:
> Been occupied doing other absurd things like DFTs of glyphs

Interesting. I looked at that before as well. Didn't get anywhere useful. I know several people did in the context of kerning. Raph being one of those.

n8willis commented:
> Interesting. I looked at that before as well. Didn't get anywhere useful. I know several people did in the context of kerning. Raph being one of those.

Yeah, it doesn't really do anything practical (more importantly, nothing that can't also be done through simpler means). More of a way of thinking about space than finding it. But I do need to spend pages saying that and showing how it compares to other concepts that people talk about but are more practical.
